feature/idCardToKotlin -> dev

aarowhead opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)
<!-- Reviewable:end -->

romrell4 commented:
Reviewed 4 of 16 files at r1.
Review status: 4 of 16 files reviewed, 1 unresolved discussion (waiting on @aarowhead)

---

*[app/src/main/java/edu/byu/suite/features/idCard/model/IdCardWrapper.kt, line 12 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF-ViDY9eO_-GQHdN8Q:-LF-ViDY9eO_-GQHdN8R:b-ov7102) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/model/IdCardWrapper.kt#L12)):*
> ```Kotlin
> 		val idCardService: IdCardService? = null
> ) {
> 	inner class IdCardService {
> ```

can this be a data class?

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->

aarowhead commented:
Review status: 11 of 18 files reviewed, 17 unresolved discussions (waiting on @romrell4 and @aarowhead)

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 28 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF30ODK-HciwZbIh6sQ:-LF3CXdf5UYqEOPs8J7a:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L28)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Can these be put into the IdCardDrawable? It might give you errors, if so, just make it not an inner class. If that gives you errors, make them non-static (but still private).
</blockquote></details>

Done.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 49 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF2zQpbC_tK3KFRyDMI:-LF3EENT2Ui5anCIU7vp:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L49)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Use new ByuSuccessCallback (ask Geoff about it)
</blockquote></details>

Done.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 68 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF2ziu60YwYC4gz4bSO:-LF3EId77S3MdtnTnB71:brzv6bf) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L68)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

What is PhotoTarget? Don't you normally `into` an ImageView? I don't think I've seen this before...
</blockquote></details>

So because we are using this bitmap to draw to a canvas, we don't have an ImageView to load it into. We need to store the two bitmaps as bitmaps and then draw them on a canvas. In Glide, you always do ```into(Target)```, ImageView is a target but because we need to do something different with the bitmap we create a custom target.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 69 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF3-tH-7q3JjrGAU-qO:-LF3HsYM9c-k9MfhwOcl:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L69)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Since we're going to have to define it ourselves anyways, let's make this a custom error. Something like "There was an issue loading the ID card photo from your account. Please try again. If the error persists, please submit feedback or contact the OIT Service Desk."
</blockquote></details>

Done.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 73 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF2zWNOAy2lf4Y5UCUS:-LF3EI-c2FxaKCWFfRfk:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L73)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Use ByuSuccessCallback
</blockquote></details>

Done.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 80 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF3-3YjDqetb0jro-bG:-LF3EtJy6HC6CfLt5EsC:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L80)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

private
</blockquote></details>

Done.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 89 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF3-JOWBAwzkoD_9bZk:-LF3IAkl0K95dE6ix-Mz:b-wqcyk2) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L89)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

What's the `+` do here?
</blockquote></details>

Apparently it isn't necessary.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 96 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF3-Pf62TVyYY5pRbSN:-LF3F0d0EqxZyRn38aSP:b-66pueq) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L96)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Whoa... target inception? Explain to me why you did this :)
</blockquote></details>

So, we need to load the bitmap for the card and once we have it we need to load the bitmap for the photo and then once we have that we have to create the drawable. So the api loads the photo into the PhotoTarget which then loads the ID card into the ByuCardTarget which then creates a IdCardDrawable object.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 118 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF30pZzEjkQvwa1NCQC:-LF3LSwHBsGQ_P6XygRh:b-7ijk85) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L118)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

These used to be hard coded in here. Is there any chance that this will return the wrong dimensions?
</blockquote></details>

I think that that is a concern. I've changed them back.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 124 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF31A0F4hIBh8QVBK0U:-LF3GJUnBKkXR5sgvleC:b5qcv9k) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L124)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Does this work on larger/smaller devices? How does it work if we're hard coding the location of the image?
</blockquote></details>

I've tested it on our small device and it works.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 134 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF31WQC9gOU9b6HBJd-:-LF3GxBX1bEp_2K8JdAF:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L134)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Can we rename `s` to a better name?
</blockquote></details>

Done.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 139 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF31aGZ1bGBqTdQpPZ3:-LF3Jt4C4zT7cMtYDinl:bbe6mr7) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L139)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Can we put this in strings.xml with a format?
</blockquote></details>

So if we make the class private then it doesn't have access the the IdCardActivity and cannot access the resource files. My thought was that this string will never need to be translated to different languages.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 143 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF31i2_1C_iBQLc1lCU:-LF3H5z2-lezhm8TCCb5:b-nv0qv3) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L143)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Do these still need to be defined? Or was that just a java thing?
</blockquote></details>

They still need to be defined to extend drawable.

---

*[app/src/main/java/edu/byu/suite/features/idCard/model/IdCardWrapper.kt, line 12 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF-ViDY9eO_-GQHdN8Q:-LF35MHFAydoDQxVt6dR:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/model/IdCardWrapper.kt#L12)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Also, put the parameter in the default constructor (like you did on the wrapper)
</blockquote></details>

Done.

---

*[app/src/main/java/edu/byu/suite/features/idCard/service/IdCardClient.kt, line 23 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF32Qql2tl880tMDcZW:-LF38OJABonRA3ZadSMe:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/service/IdCardClient.kt#L23)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Remove `: Class<Api> ` (also check all other files)
</blockquote></details>

Done.

---

*[app/src/main/res/values/strings.xml, line 169 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF32fCb34pjGiBRLnFZ:-LF3A4IKC0kXr9-g5PHB:b-sl7qqi) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/res/values/strings.xml#L169)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Isn't this already defined somewhere?
</blockquote></details>

I couldn't find it.

---

*[app/src/main/res/values/colors.xml, line 14 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF32_t3BwlzvxwXOv2D:-LF3KFdUCOxAu0jpZ4ZF:b-8cvvdr) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/res/values/colors.xml#L14)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Could we use the regular gray? Or does this look different/better?
</blockquote></details>

So we made the IdCardDrawable class private which made it so that it cannot access this color resource. If I don't set the color it makes the text black which looks good and matches the ID card better.

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed, 5 unresolved discussions (waiting on @romrell4 and @aarowhead)

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 80 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF3-3YjDqetb0jro-bG:-LF4-flyEy2nFRPOaR_d:bfhqmln) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L80)):*
<details><summary><i>Previously, aarowhead (Aaron Hill) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Not done?

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 89 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF3-JOWBAwzkoD_9bZk:-LF4-hjb7HngT0uYpBx5:baq7ikl) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L89)):*
<details><summary><i>Previously, aarowhead (Aaron Hill) wrote…</i></summary><blockquote>

Apparently it isn't necessary.
</blockquote></details>

Oh good. I'm still curious why that was valid syntax... And that it was working for you... Any ideas what it did?

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 112 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF4010L-9YJ9ebWF69L:-LF4010L-9YJ9ebWF69M:b-wxrl6t) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L112)):*
> ```Kotlin
> 			// Setting the font, color and the size of the text to be written on the ID-Card
> 			val paint = Paint()
> 			paint.color = ContextCompat.getColor(this@IdCardActivity, R.color.id_card_text_gray)
> ```

Was this unneeded? Did black look fine?

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 118 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF30pZzEjkQvwa1NCQC:-LF4-vrL1-LnyAbr5eHW:bdj50x0) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L118)):*
<details><summary><i>Previously, aarowhead (Aaron Hill) wrote…</i></summary><blockquote>

I think that that is a concern. I've changed them back.
</blockquote></details>

Even this is hard coding though, right? How does this work on larger (or smaller) devices?

---

*[app/src/main/res/values/strings.xml, line 169 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF32fCb34pjGiBRLnFZ:-LF40TGa5XBM7z6PjxcY:bs1d8n3) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/res/values/strings.xml#L169)):*
<details><summary><i>Previously, aarowhead (Aaron Hill) wrote…</i></summary><blockquote>

I couldn't find it.
</blockquote></details>

What happens in other features when personal information isn't loaded? Or are we just not checking? This feels like an error that wouldn't just pertain to ID Card...

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->

aarowhead commented:
Review status: 15 of 20 files reviewed, 5 unresolved discussions (waiting on @romrell4 and @aarowhead)

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 80 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF3-3YjDqetb0jro-bG:-LFKF6ph0WVyhtaaqlX2:bdv89gc) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L80)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Not done?
</blockquote></details>

Not applicable anymore.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 89 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF3-JOWBAwzkoD_9bZk:-LFKF9dZ6tTah36Ujhee:bqugq5j) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L89)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Oh good. I'm still curious why that was valid syntax... And that it was working for you... Any ideas what it did?
</blockquote></details>

This is all I could find on it: https://stackoverflow.com/questions/25572647/android-openrawresource-not-working-for-a-drawable

I still don't really understand it but we aren't opening raw resources anymore so it doesn't really apply.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 112 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF4010L-9YJ9ebWF69L:-LF4IbNvAEwwVcTiULzd:b-w1fqe7) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L112)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Was this unneeded? Did black look fine?
</blockquote></details>

I think black looks good, and it matches the actual ID card better.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 118 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF30pZzEjkQvwa1NCQC:-LF4J8GS5pv9jdfqaUcw:b-xmclx2) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L118)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Even this is hard coding though, right? How does this work on larger (or smaller) devices?
</blockquote></details>

We talked about this. I am going to try to build it with a constraint layout.

---

*[app/src/main/res/values/strings.xml, line 169 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF32fCb34pjGiBRLnFZ:-LFKGDSL8nsiMNAfwovH:b-ppdigd) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/res/values/strings.xml#L169)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

What happens in other features when personal information isn't loaded? Or are we just not checking? This feels like an error that wouldn't just pertain to ID Card...
</blockquote></details>

The only other place that has a similar error is on the login but it is still a different string because it says "Logged in, unable to obtain personal information..."

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 5 of 6 files at r3.
Review status: all files reviewed, 9 unresolved discussions (waiting on @romrell4 and @aarowhead)

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 53 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKiUaX0eFQagWLLhSm:-LFKiUaX0eFQagWLLhSn:b-bkzma9) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L53)):*
> ```Kotlin
> 								override fun onException(e: Exception?, model: ByteArray?, target: Target<GlideDrawable>?, isFirstResource: Boolean): Boolean {
> 									dismissProgressDialog()
> 									showErrorDialog(resources.getString(R.string.error_loading_id_photo))
> ```

So I understand correctly, this is AFTER the API loads it successfully... it could still have an error?

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 54 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKigKa-a6zu58EVVtp:-LFKigKa-a6zu58EVVtq:b-plbec6) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L54)):*
> ```Kotlin
> 									dismissProgressDialog()
> 									showErrorDialog(resources.getString(R.string.error_loading_id_photo))
> 									return false
> ```

What does the return do for us here? Does that say that we handled the error already?

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 58 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKivcj2xB4YHYCQsUp:-LFKivcj2xB4YHYCQsUq:b-1abm77) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L58)):*
> ```Kotlin
> 
> 								override fun onResourceReady(resource: GlideDrawable?, model: ByteArray?, target: Target<GlideDrawable>?, isFromMemoryCache: Boolean, isFirstResource: Boolean): Boolean {
> 									data.cardName?.let { person_name.text = it }
> ```

Can't you set the texts with null values? I don't see the need for the lets...
```kotlin
person_name.text = data.cardName
role_description.text = data.cardRole
id_number.text = data.byuId
...
...
```

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 64 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKjFTA5--1on2Q888f:-LFKjFTA5--1on2Q888g:bwoypvv) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L64)):*
> ```Kotlin
> 									data.expirationDate?.let { exp_date.text = String.format(getString(R.string.expiration_date_prefix), it) }
> 									if (data.isBeard == true) {
> 										beard_icon.visibility = View.VISIBLE
> ```

Oh wait, is the beard icon not baked into the background anymore? That's super nice...

Can't you use ternary here?

```kotlin
beard_icon.visibility = if (data.isBeard == true) View.VISIBLE else View.GONE
```

That would make it one line, and if this function runs again with isBeard being false, it would make it disappear.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 71 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKjfmS6VlFxOBkqT7S:-LFKjfmS6VlFxOBkqT7T:b-r8tg5x) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L71)):*
> ```Kotlin
> 								}
> 							})
> 							.dontAnimate()
> ```

What would it be animating?

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 73 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKjmE62RW2QMamuztP:-LFKjmE62RW2QMamuztQ:b8j4xxy) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L73)):*
> ```Kotlin
> 							.dontAnimate()
> 							.into(person_photo)
> 				} ?: onFailure(call, ByuError(getString(R.string.error_loading_id_photo)))
> ```

I've seen this string used a couple times. Can we just pass that in as the default error message to the ByuSuccessCallback, and then use this@ByuSuccessCallback to get it out? That would also standardize all of the errors

---

*[app/src/main/res/layout/id_card_activity.xml, line 2 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKk0m490Ang_7vkJQd:-LFKk0m490Ang_7vkJQe:b-5luqgx) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/res/layout/id_card_activity.xml#L2)):*
> ```XML
> <?xml version="1.0" encoding="utf-8"?>
> <android.support.constraint.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
> ```

Teach me what you did here. Walk me through it all, and why you thought it was kinda dirty at the end.

---

*[app/src/main/res/values/colors.xml, line 4 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKkBuK3pBFHAhw5Ezv:-LFKkBuK3pBFHAhw5Ezw:b-hayndu) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/res/values/colors.xml#L4)):*
> ```XML
> <resources>
>     <color name="white">#fff</color>
>     <color name="black">#000000</color>
> ```

Any reason for this? Either revert it, or do the same to white

---

*[app/src/main/res/values/strings.xml, line 169 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF32fCb34pjGiBRLnFZ:-LFKkIlT6kKwSdVlykni:bxoww72) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/res/values/strings.xml#L169)):*
<details><summary><i>Previously, aarowhead (Aaron Hill) wrote…</i></summary><blockquote>

The only other place that has a similar error is on the login but it is still a different string because it says "Logged in, unable to obtain personal information..."
</blockquote></details>

So other features that use the BYU_ID just don't check? Talk to me if you have questions

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->

aarowhead commented:
Review status: 17 of 19 files reviewed, 9 unresolved discussions (waiting on @romrell4 and @aarowhead)

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 53 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKiUaX0eFQagWLLhSm:-LFNmUjYErsmfrewmkrG:b-9ogc4n) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L53)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

So I understand correctly, this is AFTER the API loads it successfully... it could still have an error?
</blockquote></details>

I've changed it so that it is more useful. Now if the response or anything in the response that we need is null it calls onException.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 54 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKigKa-a6zu58EVVtp:-LFNffD36J8rdYIytSed:b-efjtv7) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L54)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

What does the return do for us here? Does that say that we handled the error already?
</blockquote></details>

If we want Glide to handle updating the target, we return false, if we have already handled updating the target, then we return true.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 58 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKivcj2xB4YHYCQsUp:-LFNhKMwBMfJRrB7tA_X:b-wyrg07) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L58)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Can't you set the texts with null values? I don't see the need for the lets...
```kotlin
person_name.text = data.cardName
role_description.text = data.cardRole
id_number.text = data.byuId
...
...
```
</blockquote></details>

You're right. On the String.format you need the let or else it concatenates the string "null" with the output. I had done testing with that one and saw that so I just put lets on everything without realizing it only mattered when concatenating. Its fixed now.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 64 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKjFTA5--1on2Q888f:-LFNhxnu2ps5d52F9u0a:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L64)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Oh wait, is the beard icon not baked into the background anymore? That's super nice...

Can't you use ternary here?

```kotlin
beard_icon.visibility = if (data.isBeard == true) View.VISIBLE else View.GONE
```

That would make it one line, and if this function runs again with isBeard being false, it would make it disappear.
</blockquote></details>

Done.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 71 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKjfmS6VlFxOBkqT7S:-LFNhykEAmMpkpOuBpkf:b-naj3k2) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L71)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

What would it be animating?
</blockquote></details>

Glide fades the image in to the image view unless you turn it off. I turned it off because it didn't look good in this situation. We don't really want them to see the app drawing the ID card.

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 73 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKjmE62RW2QMamuztP:-LFNwRVz7IGVM-GrKk-n:b-d4ktoc) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L73)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I've seen this string used a couple times. Can we just pass that in as the default error message to the ByuSuccessCallback, and then use this@ByuSuccessCallback to get it out? That would also standardize all of the errors
</blockquote></details>

It's only used once now.

---

*[app/src/main/res/layout/id_card_activity.xml, line 2 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKk0m490Ang_7vkJQd:-LFNiagC3DvJwhkqAGSZ:b-699g83) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/res/layout/id_card_activity.xml#L2)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Teach me what you did here. Walk me through it all, and why you thought it was kinda dirty at the end.
</blockquote></details>

I think it actually turned out pretty nice. I definitely think this is better than what we were doing before.

---

*[app/src/main/res/values/strings.xml, line 169 at r1](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LF32fCb34pjGiBRLnFZ:-LFO-9Pk8Q3Y8ierKxAs:btkrx4o) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/0ab6e8b4d64ab901e117eb5eb3f45203acf6addb/app/src/main/res/values/strings.xml#L169)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

So other features that use the BYU_ID just don't check? Talk to me if you have questions
</blockquote></details>

We talked and I moved it to an "Authorization section"

---

*[app/src/main/res/values/colors.xml, line 4 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKkBuK3pBFHAhw5Ezv:-LFNixGEBJEcN3nRKma2:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/res/values/colors.xml#L4)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Any reason for this? Either revert it, or do the same to white
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed, 1 unresolved discussion (waiting on @romrell4)

---

*[app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt, line 54 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKigKa-a6zu58EVVtp:-LFO-r281BHGXjEf08yB:b-tm04) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/java/edu/byu/suite/features/idCard/controller/IdCardActivity.kt#L54)):*
<details><summary><i>Previously, aarowhead (Aaron Hill) wrote…</i></summary><blockquote>

If we want Glide to handle updating the target, we return false, if we have already handled updating the target, then we return true.
</blockquote></details>

Awesome. Add that as a comment

---

*[app/src/main/res/layout/id_card_activity.xml, line 2 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKk0m490Ang_7vkJQd:-LFO045q-9qtZRwjTe7M:b-agixit) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/res/layout/id_card_activity.xml#L2)):*
<details><summary><i>Previously, aarowhead (Aaron Hill) wrote…</i></summary><blockquote>

I think it actually turned out pretty nice. I definitely think this is better than what we were doing before.
</blockquote></details>

Awesome. When this is the last comment, show me how you did it :)

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->

aarowhead commented:
Review status: 17 of 19 files reviewed, 1 unresolved discussion (waiting on @romrell4)

---

*[app/src/main/res/layout/id_card_activity.xml, line 2 at r3](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568#-LFKk0m490Ang_7vkJQd:-LFO4RDi3-QGFqb49f0k:b-896fix) ([raw file](https://github.com/byu-oit/android-byu-suite-v2/blob/bad7b854a190c74de53ab4233ba4c156ceb5c2fc/app/src/main/res/layout/id_card_activity.xml#L2)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Awesome. When this is the last comment, show me how you did it :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 2 of 2 files at r5.
Review status: :shipit: complete! all files reviewed, all discussions resolved

---



*Comments from [Reviewable](https://reviewable.io/reviews/byu-oit/android-byu-suite-v2/568)*
<!-- Sent from Reviewable.io -->
